include/vctr/Expressions/ReductionExpression.h

```diff
@@ -0,0 +1,184 @@
+/*
+  ==============================================================================
+    DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+
+    Copyright 2022- by sonible GmbH.
+
+    This file is part of VCTR - Versatile Container Templates Reconceptualized.
+
+    VCTR is free software: you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License version 3
+    only, as published by the Free Software Foundation.
+
+    VCTR is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+    GNU Lesser General Public License version 3 for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    version 3 along with VCTR.  If not, see <https://www.gnu.org/licenses/>.
+  ==============================================================================
+*/
+
+namespace vctr
+{
+
+class ReductionExpression
+{
+public:
+    template <is::reductionExpression Expression>
+    VCTR_FORCEDINLINE static auto reduce (const Expression& e)
+    {
+        const auto s = e.size();
+
+        // Reducing an empty vector can lead to undefined edge cases
+        VCTR_ASSERT (s > 0);
+
+        if constexpr (has::reduceVectorOp<Expression, ValueType<Expression>>)
+        {
+            return e.reduceVectorOp();
+        }
+
+        if constexpr (has::reduceNeonRegisterWise<Expression, ValueType<Expression>>)
+        {
+            return reduceNeon (e);
+        }
+
+        if constexpr (has::reduceAVXRegisterWise<Expression, ValueType<Expression>>)
+        {
+            if constexpr (is::floatNumber<ValueType<Expression>>)
+            {
+                if (Config::supportsAVX)
+                    return reduceAVX (e);
+            }
+            else
+            {
+                if (Config::supportsAVX2)
+                    return reduceAVX2 (e);
+            }
```

Comment:
In the logic above, why does the case `is::floatNumber` check for AVX and else checks for AVX2?

Reply:
Because AVX2 supplies the integer instructions for AVX registers while AVX(1) supplies the floating point instructions